### PR TITLE
Review post page and full size toc

### PR DESCRIPTION
**Pull request by b0b3rt**
This PR implements two major things:
- a new "splash" header for posts which are review winners
- a new ToC design, where the ToC is full-height, with sections that take up vertical space proportional to how much space they take up in the post

It also has https://github.com/ForumMagnum/ForumMagnum/pull/8675 merged into it, because we needed the new ReviewWinners collection to decide when to display the new splash header on post pages (as well as actually having a field for storing the image url).

EDIT: consequently, there are some changes from #8675 that need to be cleaned up, such as reverting various modifications to `usePostsList` which turned out to be unnecessary, since we aren't using them anymore.

The changes to various shared ToC components aren't ideally-architected but we've tried to feature-gate anything that might affect anyone other than LW.  (But those are the most likely annoyances to slip through, so review with an eye for those specifically.)

New Splash Header
image

With a review comment preview opened
image

With the audio player toggled open
image

New full-height ToC (with modified title row)
image

New full-height ToC with the header pinner and hiding the title
image

Comment ToC with modified title row
image

Dark mode (with review comment preview)
image

### CODE REVIEW

File: packages/lesswrong/components/comments/CommentFrame.tsx [hunk 1]
```diff
@@ -34,9 +34,9 @@ const styles = (theme: ThemeType): JssStyles => ({
   },
   new: {
     '&&': {
-      borderLeft: `solid 5px ${theme.palette.secondary.light}`,
+      borderLeft: `solid 5px ${theme.palette.secondary.light}8c`,
       '&:hover': {
-        borderLeft: `solid 5px ${theme.palette.secondary.main}`
+        borderLeft: `solid 5px ${theme.palette.secondary.main}8c`
```
Comment (b0b3rt) [inline packages/lesswrong/components/comments/CommentFrame.tsx:39]: Idk what these changes were intended to do, or how much they affect e.g. existing comments (both on LW and on other forums).  @Discordius ?

File: packages/lesswrong/components/hooks/usePaginatedResolver.ts [hunk 2]
```diff
@@ -94,10 +95,10 @@ export const usePaginatedResolver = <
   }
 
   useEffect(() => {
-    if (results?.length && results.length >= (list?.length ?? 0)) {
+    if (results?.length && results.length >= (list?.length ?? 0) && !isEqual(list, results)) {
```
Comment (b0b3rt) [inline packages/lesswrong/components/hooks/usePaginatedResolver.ts:98]: This change was motivated by an early attempt to use `usePostsList`, but even without that, I think it's still an improvement (since it now no longer causes unnecessary renders on e.g. a "load more" which doesn't actually end up pulling more records).

File: packages/lesswrong/components/posts/PostsPage/PostBodyPrefix.tsx [hunk 3]
```diff
@@ -121,7 +121,7 @@ const PostBodyPrefix = ({post, query, classes}: {
         <Info className={classes.infoIcon}/>
       </LWTooltip>
     </div>}
-    <LinkPostMessage post={post} negativeTopMargin />
+    <LinkPostMessage post={post} />
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/PostsPage/PostBodyPrefix.tsx:124]: I don't know exactly what motivated this change and whether it should be kept (and if, whether it should be forum-gated).  @Discordius ?

File: packages/lesswrong/components/posts/PostsPage/ImageCropPreview.tsx [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+  saveAllCoordinates: {
+    position: 'absolute',
+    display: 'block',
+    bottom: 0,
+    left: 0,
+    width: '100%',
+    cursor: 'pointer',
+    padding: '2px 5px',
+    userSelect: 'none', // Prevent text selection
+    color: 'black',
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/PostsPage/ImageCropPreview.tsx:122]: Will need to clean up all of the styles using absolute color values here to use `theme.palette` values (including, I think, in places like `border: '1px solid #000'`)

File: packages/lesswrong/components/posts/PostsPage/ImageCropPreview.tsx [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+      {isBoxVisible && selectedImageInfo && selectedImageInfo._id && (
+        <div className={classes.moveableBox}
+          style={moveableBoxStyle}
+          onMouseDown={handleMouseDown}>
+          <div style={boxSubContainers}>
+            <ImagePreviewSubset subBoxPosition='left' {...previewSubsetProps} />
+            <ImagePreviewSubset subBoxPosition='middle' {...previewSubsetProps} />
+            <ImagePreviewSubset subBoxPosition='right' {...previewSubsetProps} />
+          </div>
+          <div className={classes.saveAllCoordinates} onClick={saveAllCoordinates} style={{backgroundColor: showSaveAllButton ? 'rgba(0, 255, 0, 0.7)' : 'rgba(255, 0, 0, 0.3)'}}>
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/PostsPage/ImageCropPreview.tsx:426]: This should probably be a proper conditional style via `classNames`

File: packages/lesswrong/components/posts/PostsPage/ImageCropPreview.tsx [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+            {!showSaveAllButton
+              ? (<div> Must set all placements before you can save them </div>)
+              : loading
+                ? <div>Saving all placements...</div> 
+                : <div onClick={saveAllCoordinates}>{`Save all placements`}</div>
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/PostsPage/ImageCropPreview.tsx:431]: No nested ternaries, plz - extract to some more legible logic in the component that assigns to a variable that gets used here

File: packages/lesswrong/components/posts/PostsPage/PostsPagePostHeader.tsx [hunk 7]
```diff
@@ -436,13 +432,9 @@ const PostsPagePostHeader = ({post, answers = [], dialogueResponses = [], showEm
       <div className={classes.tagSection}>
         {!post.shortform && !post.isEvent && !hideTags && 
         <AnalyticsContext pageSectionContext="tagHeader">
-          <FooterTagList post={post} hideScore allowTruncate overrideMargins={true}/>
+          <FooterTagList post={post} hideScore allowTruncate overrideMargins={true} annualReviewMarketInfo={annualReviewMarketInfo} />
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/PostsPage/PostsPagePostHeader.tsx:435]: My current guess is that this change should be reverted, though I don't know for sure why it was implemented here (it's in the regular post header, not the splash page post header, so it doesn't seem related to this PR).

File: packages/lesswrong/components/posts/PostsPage/SplashHeaderImageOptions.tsx [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    {images?.map((image, index) => (
+      <div
+        className={classes.imageContainer}
+        key={index}
+        onClick={() => setImageInfo(image)}
+      >
+        <img
+          src={image.splashArtImageUrl}
+          alt={`Selectable ${index}`}
+          style={{ maxWidth: '100%', height: 'auto' }}
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/PostsPage/SplashHeaderImageOptions.tsx:47]: This should be using `classes`

File: packages/lesswrong/components/posts/PostsPage/PostsPagePostHeader.tsx (hunk 7 above)
Comment (b0b3rt) [reply packages/lesswrong/components/posts/PostsPage/PostsPagePostHeader.tsx:435]: (@rjmk)

File: packages/lesswrong/components/posts/PostsPage/PostsPageSplashHeader.tsx [hunk 9]
```diff
@@ -0,0 +1,572 @@
...
+      transition: 'opacity 0.5s',
+      transitionTimingFunction: 'ease-in',
+    },
+  },
+  centralSection: {
+    textAlign: 'center',
+    display: 'flex',
+    flexDirection: 'column',
+    position: 'relative',
+    color: 'white',
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/PostsPage/PostsPageSplashHeader.tsx:116]: Can't use absolute colors, need to use `theme.palette...`.  Not sure what this style in particular is doing and whether it should be responsive to dark mode or not.

Comment (b0b3rt) [reply packages/lesswrong/components/posts/PostsPage/PostsPageSplashHeader.tsx:116]: (will only leave this comment once per file, but using absolute color values like this fails unit tests so all instances need to be fixed.)

File: packages/lesswrong/components/posts/PostsPage/PostsPageSplashHeader.tsx [hunk 10]
```diff
@@ -0,0 +1,572 @@
...
+  const votingSystem = getVotingSystemByName(post.votingSystem ?? 'default');
+  const postActionsButton = <PostActionsButton post={post} className={classes.postActionsButton} flip />;
+
+  const topLeftSection = (
+    <div className={classes.leftSection}>
+      <Link className={classes.reviewNavigation} to="/bestoflesswrong?sort=year">
+        Ranked #{post.reviewWinner.reviewRanking + 1} of {post.reviewWinner.competitorCount} posts in the {post.reviewWinner.reviewYear} Review
+      </Link>
+      <Link className={classes.reviewNavigationMobile} to="/bestoflesswrong?sort=year">
+        #{post.reviewWinner.reviewRanking + 1} in 2021 Review
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/PostsPage/PostsPageSplashHeader.tsx:449]: I think the year here needs to come from the `reviewWinner` as well?

File: packages/lesswrong/components/tagging/FooterTag.tsx [hunk 11]
```diff
@@ -14,11 +14,9 @@ import type { TagsTooltipPreviewWrapper } from './TagsTooltip';
 const useExperimentalTagStyleSetting = new DatabasePublicSetting<boolean>('useExperimentalTagStyle', false)
 
 export const tagStyle = (theme: ThemeType): JssStyles => ({
-  marginRight: 3,
   padding: 5,
   paddingLeft: 6,
   paddingRight: 6,
-  marginBottom: 8,
```
Comment (b0b3rt) [inline packages/lesswrong/components/tagging/FooterTag.tsx:21]: Not sure why these were removed and if this was supposed to be the case for everything using `tagStyle` (both within a single forum and across forums).  @Discordius ?

File: packages/lesswrong/components/tagging/FooterTagList.tsx [hunk 12]
```diff
@@ -16,14 +16,18 @@ import { isLWorAF, taggingNamePluralSetting } from '../../lib/instanceSettings';
 import stringify from 'json-stringify-deterministic';
 import { isFriendlyUI } from '../../themes/forumTheme';
 import { FRIENDLY_HOVER_OVER_WIDTH } from '../common/FriendlyHoverOver';
+import { AnnualReviewMarketInfo, highlightMarket } from '../../lib/annualReviewMarkets';
 
 const styles = (theme: ThemeType) => ({
   root: {
-    marginTop: 8,
-    marginBottom: 8,
+    display: 'flex',
```
Comment (b0b3rt) [inline packages/lesswrong/components/tagging/FooterTagList.tsx:23]: I _think_ that all the styling changes here were @Discordius.  Same q as above - was this intended to apply for all uses of `FooterTagList` on LW, and if so, do we want to forum-gate them?

File: packages/lesswrong/lib/helpers.ts [hunk 13]
```diff
@@ -191,3 +191,9 @@ export const setAtPath = <T extends {}, V extends AnyBecauseHard>(
   }
   return value;
 }
+
+/*
+* @deprecated Only use for debugging, or else litter the console with logs
+*/
+// eslint-disable-next-line no-console
+export const trace = <A>(msg: string) => (a: A): A => { console.log(msg, a); return a }
```
Comment (b0b3rt) [inline packages/lesswrong/lib/helpers.ts:199]: @rjmk this & uses of it should probably be replaced by `loggerConstructor > logger` (sorry, am aware that is probably more annoying to pipe through `then`s).

File: packages/lesswrong/server/scripts/generativeModels/coverImage.ts [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+import { getOpenAI } from '../../languageModels/languageModelIntegration.ts';
+import { sleep } from '../../../lib/utils/asyncUtils.ts';
+import shuffle from 'lodash/shuffle';
+import { trace } from '../../../lib/helpers.ts';
+import { filterNonnull } from '../../../lib/utils/typeGuardUtils.ts';
+
+const DEPLOY = false
+
+const myMidjourneyKey = myMidjourneyAPIKeySetting.get()
+if (!myMidjourneyKey) {
```
Comment (b0b3rt) [inline packages/lesswrong/server/scripts/generativeModels/coverImage.ts:20]: This will cause an error to be thrown on server start-up for anyone who doesn't have this instance setting set, which is probably not the behavior we want

File: packages/lesswrong/lib/instanceSettings.ts [hunk 15]
```diff
@@ -211,4 +211,9 @@ export const requireReviewToFrontpagePostsSetting = new PublicInstanceSetting<bo
 export const manifoldAPIKeySetting = new PublicInstanceSetting<string | null>('manifold.reviewBotKey', null, "optional")
 export const reviewUserBotSetting = new PublicInstanceSetting<string | null>('reviewBotId', null, "optional")
 
+// TODO: remove the slack settings after we finish prototyping
+export const reviewArtSlackAPIKeySetting = new PublicInstanceSetting<string | null>('slack.apiToken', null, "optional");
+export const reviewArtSlackSigningSecretSetting = new PublicInstanceSetting<string | null>('slack.signingSecret', null, "optional");
+
+export const myMidjourneyAPIKeySetting = new PublicInstanceSetting<string | null>('myMidjourney.apiKey', null, "optional");
```
Comment (b0b3rt) [inline packages/lesswrong/lib/instanceSettings.ts:218]: Deprecate these as needed

File: packages/lesswrong/lib/routes.ts [hunk 16]
```diff
@@ -914,11 +921,6 @@ const eaLwAfForumSpecificRoutes = forumSelect<Route[]>({
       getPingback: (parsedUrl) => getPostPingbackBySlug(parsedUrl, parsedUrl.params.slug),
       background: postBackground
     },
-    {
-      name: 'bookLanding',
-      path: '/books',
-      redirect: () => `/books/2018`,
-    },
```
Comment (b0b3rt) [inline packages/lesswrong/lib/routes.ts:921]: Seems plausible that deprecating this old redirect is what we want, but do want to double-check with @Discordius

**Comment by rjmk** [comment]
It now points to the old bestof page. Chatted about this with Oli previously

File: packages/lesswrong/server/migrations/20240220T211300.remove_notnull_ReviewWinnerArts.ts [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+ * - [ ] Rename this file to something more readable
+ * - [ ] Uncomment `acceptsSchemaHash` below
+ * - [ ] Run `yarn acceptmigrations` to update the accepted schema hash (running makemigrations again will also do this)
+ */
+export const acceptsSchemaHash = "5943dd3f9d3cd8c97a6ec7bc3c11384f";
+
+export const up = async ({db}: MigrationContext) => {
+  // TODO
+  await db.none(`ALTER TABLE "ReviewWinnerArts" 
+    ALTER COLUMN "postId" DROP NOT NULL`
```
Comment (b0b3rt) [inline packages/lesswrong/server/migrations/20240220T211300.remove_notnull_ReviewWinnerArts.ts:39]: Don't forget to undo & delete this migration.

File: packages/lesswrong/server/scripts/generativeModels/coverImage.ts [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+
+    return [...prevUrls, ...images]
+  }, Promise.resolve([]) as Promise<string[]>)
+}
+
+async function main () {
+  const newImages = await generateCoverImages({limit: 9999})
+  if (newImages.length === 0) return
+  await sleep(10_000)
+  await main()
```
Comment (b0b3rt) [inline packages/lesswrong/server/scripts/generativeModels/coverImage.ts:306]: ?!

File: packages/lesswrong/server/scripts/generativeModels/coverImage.ts [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+              .then(trace(`Upscaled & saved ${el}`))
+              .then(urls => postPromptImages(el, postedEssay, filterNonnull(urls)))
+            return [...ims, ...im]
+          }, Promise.resolve([]) as Promise<string[]>))
+
+    return [...prevUrls, ...images]
+  }, Promise.resolve([]) as Promise<string[]>)
+}
+
+async function main () {
```
Comment (b0b3rt) [inline packages/lesswrong/server/scripts/generativeModels/coverImage.ts:302]: This should definitely be called something other than `main` if it's being assigned to `Globals`

File: package.json [hunk 20]
```diff
@@ -323,6 +323,7 @@
     "@graphql-codegen/cli": "^1.9.1",
     "@graphql-codegen/fragment-matcher": "^1.9.0",
     "@graphql-codegen/introspection": "^1.9.0",
+    "@slack/bolt": "^3.17.1",
```
Comment (b0b3rt) [inline package.json:326]: If this is no longer being used by the script to generate images, please delete it

File: packages/lesswrong/components/sequences/TopPostsPage.tsx [hunk 21]
```diff
@@ -0,0 +1,906 @@
...
+// TODO: not actually sure we want to use default coordinates if coordinates are missing
+// And even if we do, these values are definitely wrong (i.e. need to figure out aspect ratio)
+const DEFAULT_SPLASH_ART_COORDINATES: Omit<SplashArtCoordinates, '_id' | 'reviewWinnerArtId'> = {
+  leftHeightPct: .2, leftWidthPct: .2, leftXPct: .2, leftYPct: .2, leftFlipped: false,
+  middleHeightPct: .2, middleWidthPct: .2, middleXPct: .2, middleYPct: .2, middleFlipped: false,
+  rightHeightPct: .2, rightWidthPct: .2, rightXPct: .2, rightYPct: .2, rightFlipped: false,
+};
+
+// TODO: update the description to be appropriate for this page
+const description = `All of ${siteNameWithArticleSetting.get()}'s posts, filtered and sorted however you want`;
```
Comment (b0b3rt) [inline packages/lesswrong/components/sequences/TopPostsPage.tsx:71]: I think this is used for SEO + search result purposes.  Need to figure out what goes here.  @rjmk @Discordius @darkruby501

File: packages/lesswrong/server/resolvers/reviewWinnerResolvers.ts [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+    reviewWinner
+  }));
+}
+
+defineQuery({
+  name: 'GetAllReviewWinners',
+  resultType: '[Post!]!',
+  fn: async (root, args, context) => {
+    // TODO: reinstate the cache
+    const cacheStale = true // moment(REVIEW_WINNER_CACHE.lastUpdatedAt).isBefore(moment(new Date()).subtract(1, 'hour'));
```
Comment (b0b3rt) [inline packages/lesswrong/server/resolvers/reviewWinnerResolvers.ts:31]: Don't foget to undo this!!

File: packages/lesswrong/server/migrations/20240220T211300.remove_notnull_ReviewWinnerArts.ts (hunk 17 above)
Comment (darkruby501) [reply packages/lesswrong/server/migrations/20240220T211300.remove_notnull_ReviewWinnerArts.ts:39]: Done

File: packages/lesswrong/components/posts/PostsPage/PostsPageSplashHeader.tsx (hunk 10 above)
Comment (darkruby501) [reply packages/lesswrong/components/posts/PostsPage/PostsPageSplashHeader.tsx:449]: Done

File: packages/lesswrong/components/posts/PostsPage/PostsPageSplashHeader.tsx (hunk 9 above)
Comment (darkruby501) [reply packages/lesswrong/components/posts/PostsPage/PostsPageSplashHeader.tsx:116]: Fixed for this file

**Review by rjmk — APPROVED** [review]
